File: chimera/internal/repl/repl.go
package repl

import (
	"bufio"
	"fmt"
	"goInterpreter/chimera/internal/evaluator"
	"goInterpreter/chimera/internal/lexer"
	"goInterpreter/chimera/internal/object"
	"goInterpreter/chimera/internal/parser"
	"io"
)

const PROMPT = ">> "

const CHIMERA = `
(╯°□°）╯︵ ┻━┻ 
`

func Start(in io.Reader, out io.Writer) {
	scanner := bufio.NewScanner(in)
	env := object.NewEnvironment()

	for {
		fmt.Fprint(out, PROMPT)
		scanned := scanner.Scan()

		if !scanned {
			return
		}

		line := scanner.Text()
		// Lexer Struct = readPositon = 1 , ch = first character
		l := lexer.New(line)
		p := parser.New(l)

		program := p.ParseProgram()
		if len(p.Errors()) != 0 {
			printParserErrors(out, p.Errors())
			continue
		}

		evaluated := evaluator.Eval(program, env)
		if evaluated != nil {
			io.WriteString(out, evaluated.Inspect())
			io.WriteString(out, "\n")
		}
	}
}

func printParserErrors(out io.Writer, errors []string) {
	io.WriteString(out, CHIMERA)
	io.WriteString(out, "Woops! We ran into some errors here!\n")
	io.WriteString(out, " parser errors:\n")
	for _, msg := range errors {
		io.WriteString(out, "\t"+msg+"\n")
	}
}

File: chimera/src/repl/repl.rs
use std::io::{self, Write};
use crate::evaluator::evaluator;
use crate::lexer::lexer::Lexer;
use crate::object::object::{Environment, Object};
use crate::parser::parser;

pub const PROMPT: &'static str = ">> ";
pub const CHIMERA: &'static str = r#"
(╯°□°）╯︵ ┻━┻ 
"#;

pub fn start_repl() {
    let mut environment = Environment::new();
    loop {
        let lexer = Lexer::new(prompt());
        let mut parser = parser::Parser::new(lexer);
        let program = parser.parse_program();

        if parser.errors().len() != 0 {
            print_parser_error(parser.errors());
            continue;
        };
        
        let evaluated = evaluator::eval(&program, &mut environment);
        match evaluated {
            Object::Null => (),
            _ => println!("{}", evaluated.inspect()),
        }
    }
}

fn prompt() -> String {
    let mut user_input = String::new();
    print!("{PROMPT}");

    std::io::stdout().flush().unwrap();
    io::stdin()
        .read_line(&mut user_input)
        .expect("Failed to read line");

    return user_input.trim().to_string()
}

fn print_parser_error(errors: Vec<String>) {
    println!("{}", CHIMERA);
    println!("Woops! We ran into some errors here!");
    println!("Parser errors:");
    errors.iter().for_each(
        |e| println!("\t{}", e)
    );
}
